### PR TITLE
:sparkles: Allow multiple projects open at once

### PR DESCRIPTION
**Pull request by markmehere**
This allows multiple projects to be open at the same time (useful for transferring code).

It adds a working Window menu for Mac OS X. Mixed Context Mode, which made the rest of the pull request easier, made implementing this menu slightly difficult.

The feature is a light-touch retaining the existing open functionality everywhere but adding a single button "Open new project window" in the project menu (see picture).

In addition:

* Code is added to check the same project is not opened several times (could spell trouble)
* Code is added to ensure the global notepad is updated during switches
* It incorporates the `package.json` changes from #377 because otherwise new windows would fail with an unknown `toolbar` option
* It removes `this.authoring.title` window titles for all platforms as these were not working and inconsistently applied

image

### CODE REVIEW

**Comment by FlipFloop** [comment]
I think another way of doing this would be able to generate a `./common` folder. So we can simply have that on the filesystem, add sprites, movements scripts, sounds etc. and just choose whether to import from common? Almost like a template?

**Comment by FlipFloop** [comment]
Also, can't work on this as M1 not enabled :(

**Comment by markmehere** [comment]
@FlipFloop Yes until the Apple silicon builds are merged - none of the branches are going to run natively on your computer. 🤔 You have to pull both changes into a single branch.

**Comment by CosmoMyzrailGorynych** [comment]
Wow @markmehere, you were busy these two weeks! I returned from my business trip just yesterday and will need time to merge all this :eyes:

File: src/riotTags/main-menu/main-menu-project.tag [hunk 1]
```diff
@@ -121,6 +124,22 @@ main-menu-project
...
+                icon: 'ct_ide.png',
+                frame: true,
+                width: 1024,
+                height: 720,
+                position: 'center',
+                min_width: 800,
+                min_height: 600,
+                resizable: true,
+                fullscreen: false,
+                title: 'ctjs'
```
Comment (CosmoMyzrailGorynych) [inline src/riotTags/main-menu/main-menu-project.tag]: These all can be replaced with `require('./package.json').window`

Comment (CosmoMyzrailGorynych) [reply src/riotTags/main-menu/main-menu-project.tag]: `min_width` and `min_height` are incorrect; when needed, on tiny screens, ct.js can be zoomed out.

**Review by CosmoMyzrailGorynych — CHANGES_REQUESTED** [review]
Window creation should follow what is in package.json of the app.

File: src/riotTags/main-menu/main-menu-project.tag (hunk 1 above)
Comment (markmehere) [reply src/riotTags/main-menu/main-menu-project.tag]: Mac OS X defaults to these minimum values at random times. This is a nw.js issue. But rest sounds good to me.

**Comment by FlipFloop** [comment]
> I think another way of doing this would be able to generate a `./common` folder. So we can simply have that on the filesystem, add sprites, movements scripts, sounds etc. and just choose whether to import from common? Almost like a template?

Is this idea dumb? I think it'd save time from copying stuff over and over.

**Comment by CosmoMyzrailGorynych** [comment]
> Is this idea dumb?

It's as raw as a living chicken. Too broad, too vague, and not relying on existing functionality.

**Comment by FlipFloop** [comment]
> > Is this idea dumb?
> 
> It's as raw as a living chicken. Too broad, too vague, and not relying on existing functionality.

Solid. I'll **cook** it for us all then.

**Comment by markmehere** [comment]
@FlipFloop I like the idea of a game component library and I think it's not the first time it's been raised.

Though my PRs seem all over the place, almost all are based on issues I'm encountering while making games.

So this feature is simply because I need to bring data across from one project to another. Doing it without the multi-windows is painful.

This is not, for example, a solution for https://github.com/ct-js/ct-js/issues/48 - which is slightly different.

A component library for ct.js game developers (so you can drag in a timer or quit button or flying cat sprite) can be easy to do badly. So the trick is to get a really crisp implementation that's as simple as possible.

I'm sort of thinking of an online GitHub repository that can inject code and assets anywhere in the ct.js project. The GitHub would make it a walled garden. The idea is once you imported that flying cat from the online repository it would become like any other asset in your project. You could modify it however you want.

In terms of priority, I probably would like to see drag-n-drop text sprites first. It just seems like a really big gap at the moment.

**Comment by markmehere** [comment]
This is still in draft, btw, because without a proper Windows menu - the multi-window on Mac OS X feels really strange. Apple causing trouble again.

**Comment by markmehere** [comment]
Every other pull request though is good to go. #372 👀 for example. This is, IMHO, important for OS X.

**Comment by CosmoMyzrailGorynych** [comment]
@markmehere https://github.com/ct-js/ct-js/issues/203 wink-wink several base classes + shared behaviors wink-wink

**Comment by CosmoMyzrailGorynych** [comment]
> This is still in draft, btw, because without a proper Windows menu - the multi-window on Mac OS X feels really strange. Apple causing trouble again. 🤬

AFAIK the Window menu can be replaced with nw.js API: https://nwjs.readthedocs.io/en/latest/References/Menu/#menucreatemacbuiltinappname-options-mac

**Comment by markmehere** [comment]
> AFAIK the Window menu can be replaced with nw.js API: https://nwjs.readthedocs.io/en/latest/References/Menu/#menucreatemacbuiltinappname-options-mac

This sounds good and I spotted it too - but actually isn't anywhere near as useful as it could be and (for document management) the heavy lifting is left entirely to the app.

File: src/js/utils/windowMenuHelper.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+(function windowMenuHelper() {
+  const updateWindowMenu = async () => {
+    await new Promise(resolve => setTimeout(() => resolve(), 300));
+    const allWindows = await (new Promise((resolve) => {
+      nw.Window.getAll((windows) => {
+        resolve(windows.map(win => {
```
Comment (CosmoMyzrailGorynych) [inline src/js/utils/windowMenuHelper.js:6]: Cool! Now I vaguely know what it is all about! 😅

**Review by CosmoMyzrailGorynych — APPROVED** [review]

**Pull request merged**